Node.stack_trace should have innermost frame last

SherlockNoMad opened the pull request:
Both fx.Tracer and Dynamo should store node.stack_trace in the "innermost frame last" order. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95592



cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

SherlockNoMad committed:
Fode.sstack_trace should have innermost frame last

[ghstack-poisoned]

ezyang commented on torch/fx/proxy.py:
```diff
@@ -180,9 +180,9 @@ def create_proxy(self, kind: str, target: Target, args: Tuple[Any, ...], kwargs:
             user_frame = self._find_user_frame()
             if user_frame:
-                walk_stack_gen = traceback.walk_stack(user_frame)
-                summary = traceback.StackSummary.extract(walk_stack_gen)  # type: ignore[arg-type]
```
interesting, I wonder why we did it this way originally

SherlockNoMad replied:
Not sure... traceback.extract_stack(frame) would reverse the stack for user.

ezyang commented on torch/fx/node.py:
```diff
@@ -363,9 +363,13 @@ def update_kwarg(self, key : str, arg : Argument) -> None:
+        When traced with dynamo, this property will be populated by default by
+        `OutputGraph.create_proxy`.
+
+        stack_trace would have the innermost frame at the bottom.
```
here "bottom" means "end of list" right

ezyang commented on torch/fx/graph.py:
```diff
@@ -451,24 +451,28 @@ def append_stacktrace_summary(node : Node):
+                        # iterate backwards to find the first line that starts
+                        # with 'File '
+                        idx = len(lines) - 1
+                        while idx >= 0:
```
nit: use a range loop here? `for idx in range(len(lines)-1, -1, -1)`

SherlockNoMad replied:
ok, I rewrote to simplify the code.

ezyang commented on torch/fx/graph.py:
```diff
@@ -451,24 +451,28 @@ def append_stacktrace_summary(node : Node):
                             if line.startswith('File '):
                                 break
-                            idx += 1
+                            idx -= 1
```
nit: You can use an else branch on the loop to elegantly set idx to None when it is not found

```
while ...:
    ...
else:
   idx = None
```

ezyang reviewed (approved):
Thanks!

SherlockNoMad committed:
Update on "Node.stack_trace should have innermost frame last"

Both fx.Tracer and Dynamo should store node.stack_trace in the "innermost frame last" order. 




cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge

SherlockNoMad commented:
@pytorchbot merge
